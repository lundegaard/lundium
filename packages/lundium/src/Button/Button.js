import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import { cx, isString, noop, valueMirror } from 'ramda-extension';

import Icon from '../Icon';
import Text from '../Text';
import Link from '../Link';

/**
 * Represents a button. Can act as a link if `href` is passed.
 */
const Button = forwardRef(
	(
		{
			kind = 'default',
			LinkComponent = Link,
			onClick = noop,
			outline = false,
			type = 'button',
			blank,
			children,
			className,
			disabled,
			icon,
			size,
			...rest
		},
		ref,
	) => {
		const C = rest.href || rest.to ? LinkComponent : 'button';
		const attrs = !rest.href && !rest.to ? { type } : {};

		return (
			<Text
				as={C}
				{...rest}
				className={cx(
					'btn',
					{
						'btn-blank': blank,
						[`btn-${size}`]: size,
						[`btn-${kind}`]: !blank && !outline,
						[`btn-outline-${kind}`]: outline,
					},
					className,
				)}
				onClick={disabled ? noop : onClick}
				disabled={disabled}
				ref={ref}
				{...attrs}
			>
				{isString(icon) ? <Icon type={icon} /> : icon}
				{children}
			</Text>
		);
	},
);

export const BUTTON_TYPES = valueMirror([
	'primary',
	'default',
	'success',
	'info',
	'warning',
	'danger',
	'link',
	'secondary',
	'blank',
]);

Button.propTypes = {
	LinkComponent: PropTypes.elementType,
	/** If `true`, than no padding */
	blank: PropTypes.bool,
	/** Children to be rendered in the main container. */
	children: PropTypes.node,
	/** Class name for root. */
	className: PropTypes.string,
	disabled: PropTypes.bool,
	/** If set, button will act like a Link. */
	href: PropTypes.string,
	/** Supporting only values from Icon component. */
	icon: PropTypes.node,
	/** Kind of button. */
	kind: PropTypes.oneOf(Object.values(BUTTON_TYPES)),
	/** Handler for `click` event. */
	onClick: PropTypes.func,
	/** Outline / ghost kind of button. */
	outline: PropTypes.bool,
	/** Size of the button. If falsy, default size is applied. */
	size: PropTypes.oneOf(['lg', 'sm', 'xs']),
	/** Button type HTML attribute. */
	type: PropTypes.string,
	...Text.propTypes,
};

export default Button;
