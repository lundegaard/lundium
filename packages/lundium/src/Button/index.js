export { default, BUTTON_TYPES } from './Button';
