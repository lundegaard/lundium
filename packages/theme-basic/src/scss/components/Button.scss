@import '../core';
@import '~bootstrap/scss/buttons';

/*
 -  Buttons
 -  (<button class="btn" />)
 */


.btn {
	min-width: 130px;
	// Offset padding to center font line height (Open sans)
	padding: logical $btn-padding-y $btn-padding-x ( $btn-padding-y + .1rem );
	display: inline-flex;
	align-items: center;
	justify-content: center;
	text-transform: uppercase;
	white-space: nowrap;

	&-sm {
		min-width: auto;
		// Offset padding to center font line height (Open sans)
		padding: logical $btn-padding-y-sm $btn-padding-x-sm ( $btn-padding-y-sm + .1rem );
	}

	&-xs {
		min-width: auto;
		padding: logical $btn-padding-y-xs $btn-padding-x-xs;
	}

	&-primary {
		@include button-hover-state($btn-primary-hover-bg-color, $btn-primary-hover-border-color, $btn-primary-hover-color);
		@include button-active-focus-state($btn-primary-focus-bg-color, $btn-primary-focus-border-color, $btn-primary-focus-color, $btn-primary-focus-box-shadow);
	}


	&-secondary {
		@include button-hover-state($btn-secondary-hover-bg-color, $btn-secondary-hover-border-color, $btn-secondary-hover-color);
		@include button-active-focus-state($btn-secondary-focus-bg-color, $btn-secondary-focus-border-color, $btn-secondary-focus-color, $btn-secondary-focus-box-shadow);
	}

	&-danger {
		@include button-hover-state($btn-danger-hover-bg-color, $btn-danger-hover-border-color, $btn-danger-hover-color);
		@include button-active-focus-state($btn-danger-focus-bg-color, $btn-danger-focus-border-color, $btn-danger-focus-color, $btn-danger-focus-box-shadow);
	}

	&-blank {
		background-color: transparent;
		border: none;
		@include button-hover-state($btn-blank-hover-bg-color, $btn-blank-hover-border-color, $btn-blank-hover-color);
		@include button-active-focus-state(transparent, transparent, $btn-blank-focus-color, none);

		&.btn.disabled,
		&.btn:disabled {
			background-color: transparent;
		}
	}

	&-outline-primary {
		@include button-hover-state($btn-outline-primary-hover-bg-color, $btn-outline-primary-hover-border-color, $btn-outline-primary-hover-color);
		@include button-active-focus-state($btn-outline-primary-focus-bg-color, $btn-outline-primary-focus-border-color, $btn-outline-primary-focus-color, $btn-outline-primary-focus-box-shadow);
	}

	&-outline-secondary {
		border-color: $theme-gray-600;
		color: $theme-gray-600;

		@include button-hover-state($btn-outline-secondary-hover-bg-color, $btn-outline-secondary-hover-border-color, $btn-outline-secondary-hover-color);
		@include button-active-focus-state($btn-outline-secondary-focus-bg-color, $btn-outline-secondary-focus-border-color, $btn-outline-secondary-focus-color, $btn-outline-secondary-focus-box-shadow);
	}

	&-outline-danger {
		@include button-hover-state($btn-outline-danger-hover-bg-color, $btn-outline-danger-hover-border-color, $btn-outline-danger-hover-color);
		@include button-active-focus-state($btn-outline-danger-focus-bg-color, $btn-outline-danger-focus-border-color, $btn-outline-danger-focus-color, $btn-outline-danger-focus-box-shadow);
	}

	&.disabled,
	&:disabled {
		color: $btn-disabled-color;
		background-color: $btn-disabled-bg-color;
		border-color: $btn-disabled-border-color;
		cursor: not-allowed;

		[class^='icon'],
		[class*='icon'] {
			// stylelint-disable-next-line
			color: $disabled-color !important;
			opacity: $disabled-opacity;
		}
	}

	[class^='icon'],
	[class*='icon'] {
		padding-inline-end: ( $spacer * .25 );
		// Color has to be important to override default color of icon
		// stylelint-disable-next-line
		color: inherit !important;

		&::before {
			// disable text decoration on icon
			display: inline-block;
		}
	}

	&--icon {
		min-width: auto;
		padding: logical $btn-padding-y;
		border-radius: 50%;

		[class^='icon'],
		[class*='icon'] {
			padding-inline-end: 0;
		}
	}
}

// Little hack to switch arrows in RTL
[dir='rtl'] {
	.icon--type-next,
	.icon--type-back {
		transform: rotate(180deg);
		padding-left: 0;
		padding-right: ( $spacer * .25 );
	}
}