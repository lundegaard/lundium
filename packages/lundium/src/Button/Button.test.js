import React from 'react';
import { shallow } from 'enzyme';

import Button from './Button';

describe('<Button />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<Button {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
