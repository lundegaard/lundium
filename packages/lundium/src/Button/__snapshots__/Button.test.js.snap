// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches snapshot when shallowly rendered 1`] = `
<forwardRef(Text)
  as="button"
  className="btn btn-default"
  onClick={[Function]}
  type="button"
/>
`;
